pkg/registry/populator_test.go

```diff
@@ -670,3 +678,173 @@ func CheckBundlesHaveContentsIfNoPath(t *testing.T, db *sql.DB) {
 		require.NotZero(t, bundlelen.Int64, "length of bundle for %s should not be zero, it has no bundle path", name.String)
 	}
 }
+
+func TestDeprecateBundle(t *testing.T) {
+	type args struct {
+		bundles []string
+	}
+	type pkgChannel map[string][]string
+	type expected struct {
+		err                  error
+		remainingBundles     []string
+		deprecatedBundles    []string
+		remainingPkgChannels pkgChannel
+	}
+	tests := []struct {
+		description string
+		args        args
+		expected    expected
+	}{
+		{
+			description: "BundleDeprecated/IgnoreIfNotInIndex",
+			args: args{
+				bundles: []string{
+					"quay.io/test/etcd.0.6.0",
+				},
+			},
+			expected: expected{
+				err: errors.NewAggregate([]error{fmt.Errorf("error deprecating bundle quay.io/test/etcd.0.6.0: %s", registry.ErrBundleImageNotInDatabase)}),
+				remainingBundles: []string{
+					"quay.io/test/etcd.0.9.0",
+					"quay.io/test/etcd.0.9.2",
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.14.0",
+					"quay.io/test/prometheus.0.15.0",
+				},
+				deprecatedBundles: []string{},
+				remainingPkgChannels: pkgChannel{
+					"etcd": []string{
+						"beta",
+						"alpha",
+						"stable",
+					},
+					"prometheus": []string{
+						"preview",
+						"stable",
+					},
+				},
+			},
+		},
+		{
+			description: "BundleDeprecated/SingleChannel",
+			args: args{
+				bundles: []string{
+					"quay.io/test/prometheus.0.15.0",
+				},
+			},
+			expected: expected{
+				err: nil,
+				remainingBundles: []string{
+					"quay.io/test/etcd.0.9.0",
+					"quay.io/test/etcd.0.9.2",
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.15.0",
+				},
+				deprecatedBundles: []string{
+					"quay.io/test/prometheus.0.15.0",
+				},
+				remainingPkgChannels: pkgChannel{
+					"etcd": []string{
+						"beta",
+						"alpha",
+						"stable",
+					},
+					"prometheus": []string{
+						"preview",
+						"stable",
+					},
+				},
+			},
+		},
+		{
+			description: "BundleDeprecated/ChannelRemoved",
+			args: args{
+				bundles: []string{
+					"quay.io/test/etcd.0.9.2",
+				},
+			},
+			expected: expected{
+				err: nil,
+				remainingBundles: []string{
+					"quay.io/test/etcd.0.9.2",
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.14.0",
+					"quay.io/test/prometheus.0.15.0",
+				},
+				deprecatedBundles: []string{
+					"quay.io/test/etcd.0.9.2",
+				},
+				remainingPkgChannels: pkgChannel{
+					"etcd": []string{
+						"alpha",
+						"stable",
+					},
+					"prometheus": []string{
+						"preview",
+						"stable",
+					},
+				},
+			},
+		},
```

Comment:
this is looking pretty good, but doesn't quite convince me that there are no bugs - at the very least I think this should verify the full end state of the graph, not just the bundles/packages/channels.

I also want to make sure that this covers:

- bundle is in two channels
- bundle is in two packages
- bundle is skipped by another bundle
- bundle has children that are skipped that get removed

(it might already cover these implicitly, just need to check)